File: typescript-datamodelpoisoning/secure9.ts
import * as ort from "onnxruntime-node";
import * as crypto from "crypto";

const EXPECTED_HASH = "a1b2c3d4e5f67890abcdef1234567890a1b2c3d4e5f67890abcdef1234567890";
const TRUSTED_URL = "https://internal.example.com/models/classifier.onnx";

async function loadVerifiedModel(): Promise<ort.InferenceSession> {
  const response = await fetch(TRUSTED_URL);
  const buffer = Buffer.from(await response.arrayBuffer());
  const hash = crypto.createHash("sha256").update(buffer).digest("hex");
  if (hash !== EXPECTED_HASH) {
    throw new Error("integrity check failed");
  }
  return ort.InferenceSession.create(buffer);
}

loadVerifiedModel();

File: go-datamodelpoisoning/secure9.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"net/http"

	ort "github.com/yalue/onnxruntime_go"
)

const (
	trustedModelURL = "https://internal.example.com/models/classifier.onnx"
	expectedSHA256  = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
)

func loadVerifiedModel() (*ort.AdvancedSession, error) {
	resp, err := http.Get(trustedModelURL)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	sum := sha256.Sum256(data)
	if hex.EncodeToString(sum[:]) != expectedSHA256 {
		return nil, fmt.Errorf("integrity check failed")
	}
	return ort.NewAdvancedSession(
		string(data),
		[]string{"input"},
		[]string{"output"},
		nil, nil, nil,
	)
}
